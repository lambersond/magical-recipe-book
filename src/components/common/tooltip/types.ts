import type { Placement } from '@floating-ui/react'

export type TooltipProps = {
  children: React.ReactNode
  title: React.ReactNode
  placement?: Placement
  asChild?: boolean
}
