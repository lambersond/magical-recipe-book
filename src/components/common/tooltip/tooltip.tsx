'use client'

import {
  cloneElement,
  createContext,
  forwardRef,
  isValidElement,
  useContext,
  useMemo,
  useState,
} from 'react'
import {
  useFloating,
  autoUpdate,
  offset,
  flip,
  shift,
  useHover,
  useFocus,
  useDismiss,
  useRole,
  useInteractions,
  useMergeRefs,
  FloatingPortal,
  type Placement,
} from '@floating-ui/react'
import type { TooltipProps } from './types'

interface TooltipOptions {
  placement?: Placement
}

function useTooltip({ placement = 'top' }: TooltipOptions = {}) {
  const [open, setOpen] = useState(false)

  const data = useFloating({
    placement,
    open,
    onOpenChange: setOpen,
    whileElementsMounted: autoUpdate,
    middleware: [
      offset(8),
      flip({
        crossAxis: placement.includes('-'),
        fallbackAxisSideDirection: 'start',
        padding: 8,
      }),
      shift({ padding: 8 }),
    ],
  })

  const context = data.context

  const hover = useHover(context, {
    move: false,
    enabled: true,
  })
  const focus = useFocus(context, {
    enabled: true,
  })
  const dismiss = useDismiss(context)
  const role = useRole(context, { role: 'tooltip' })

  const interactions = useInteractions([hover, focus, dismiss, role])

  return useMemo(
    () => ({
      open,
      setOpen,
      ...interactions,
      ...data,
    }),
    [open, setOpen, interactions, data],
  )
}

type ContextType = ReturnType<typeof useTooltip> | undefined

const TooltipContext = createContext<ContextType>(undefined)

const useTooltipContext = () => {
  const context = useContext(TooltipContext)

  return context as ReturnType<typeof useTooltip>
}

function TooltipContainer({
  children,
  ...options
}: { children: React.ReactNode } & TooltipOptions) {
  const tooltip = useTooltip(options)
  return (
    <TooltipContext.Provider value={tooltip}>
      {children}
    </TooltipContext.Provider>
  )
}

const TooltipTrigger = forwardRef<
  HTMLElement,
  React.HTMLProps<HTMLElement> & { asChild?: boolean }
>(function TooltipTrigger({ children, asChild = false, ...props }, propRef) {
  const context = useTooltipContext()
  const childrenRef = (children as any).ref
  const ref = useMergeRefs([context.refs.setReference, propRef, childrenRef])

  if (asChild && isValidElement(children)) {
    return cloneElement(
      children,
      context.getReferenceProps({
        ref,
        ...props,
        ...(children.props as Record<string, any>),
        'data-state': context.open ? 'open' : 'closed',
      } as any),
    )
  }

  return (
    <button
      ref={ref}
      className='w-fit'
      data-state={context.open ? 'open' : 'closed'}
      {...context.getReferenceProps(props)}
    >
      {children}
    </button>
  )
})

const TooltipContent = forwardRef<
  HTMLDivElement,
  React.HTMLProps<HTMLDivElement>
>(function TooltipContent({ style, ...props }, propRef) {
  const context = useTooltipContext()
  const ref = useMergeRefs([context.refs.setFloating, propRef])

  if (!context.open) return

  return (
    <FloatingPortal>
      <div
        ref={ref}
        style={{
          ...context.floatingStyles,
          ...style,
        }}
        {...context.getFloatingProps(props)}
      />
    </FloatingPortal>
  )
})

export function Tooltip({
  children,
  title,
  placement,
  asChild = false,
}: Readonly<TooltipProps>) {
  return (
    <TooltipContainer placement={placement}>
      <TooltipTrigger asChild={asChild}>{children}</TooltipTrigger>
      <TooltipContent className='tooltip max-w-[calc(100vw_- _8px)] bg-neutral-700 px-3 py-1 mb-1 rounded-lg z-10000'>
        {title}
      </TooltipContent>
    </TooltipContainer>
  )
}
