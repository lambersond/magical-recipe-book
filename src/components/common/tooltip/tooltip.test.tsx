import { render, useUser } from '@test-utils'
import { Tooltip } from './tooltip'

describe('components/common/tooltip', () => {
  it('should match the snapshot', () => {
    expect(
      render(<Tooltip title='title'>Hover</Tooltip>).asFragment(),
    ).toMatchSnapshot()
  })

  it('should match the snapshot with placement - asChild component', async () => {
    const { user, getByText } = useUser(
      <Tooltip title='title' asChild>
        <p>Hover</p>
      </Tooltip>,
    )

    await user.hover(getByText('Hover'))

    expect(getByText('title')).toBeInTheDocument()
  })

  it('should match the snapshot with placement - text chilc', async () => {
    const { user, getByText } = useUser(<Tooltip title='title'>Hover</Tooltip>)

    await user.hover(getByText('Hover'))

    expect(getByText('title')).toBeInTheDocument()
  })
})
