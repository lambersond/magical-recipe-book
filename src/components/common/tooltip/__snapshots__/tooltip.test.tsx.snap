// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/common/tooltip should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="w-fit"
    data-state="closed"
  >
    Hover
  </button>
</DocumentFragment>
`;
